ruby/examples/lib/apply_order_discount.rb

```diff
@@ -0,0 +1,44 @@
+class ApplyOrderDiscount
+  Request = Struct.new(:id)
+  Response = Struct.new(:discount, :has_discount?)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def apply(request)
+    order = gateway.find_order_by_id(request.id)
+
+    unless order.has_discount?
+      order.discount = discount_amount
+      gateway.save_order_discount(request.id, order)
+    end
+
+    Response.new(order.discount, order.has_discount?)
+  end
+
+  private
+
+  def discount_amount
+    10
+  end
+
+  class DiscountableOrder < Struct.new(:discount)
+    def has_discount?
+      discount > 0
+    end
+  end
+
+  class Gateway
```

Comment:
Unsure of the problem with this. Until we need to share Gateway's logic, I'm happy to keep it contained.

It's also quite nice to show how simple a gateway can be, doesn't even need to be in another file, just needs to wrap ActiveRecord and speak in Domain models.